Short-Answer/Algorithms_Answers.md

```diff
@@ -2,14 +2,99 @@
 
 ## Exercise I
 
-a)
+a) The snippet's line-by-line big-o looks like:
+```python
+    a = 0                             O(1)
+    while (a < n * n * n):          O(n)
+      a = a + n * n                   O(1)
+```
+Lines 7 and 9 are all individual statements which are O(1) while loops are O(however many elements there are).
 
+Multiply the body of the loops times the loops themselves:
+O(1)
+O(n * 1)
 
-b)
+Add sequences:
+O(1 + n * 1)
 
+Final Answer:
+O(1 + n)
+
+
+
+b)The snippet's line-by-line big-o looks like:
+```
+    sum = 0                           O(1)
+    for i in range(n):              O(n)
+      j = 1                           O(1)
+      while j < n:                  O(n)
+        j *= 2                        O(1)
+        sum += 1                      O(1)
+```
+Lines 15, 17, 19, and 20 are all individual statements which are O(1) while loops are O(however many elements there are). Adding up the Os of lines 19 and 20 we get O(2):
+
+O(1)
+O(n)
+  O(1)
+  O(n)
+    O(2)
+
+Multiply the body of the loops times the loops themselves except for those that have more than one item in their body:
+O(1)
+O(n * 1)
+  O(1)
+  O(n * 2)
+
+Add sequences
+O(1)
+O(n * 1)
+
+
+Multiply body of loops: 
+O(1)
+O(n * 1 * (1 + n * 2))
+
+Add sequences:
+O(1 + n * 1 * (1 + n * 2))
+
+Math: 
+1 + n * (1 + 2n)
+1 + 1n + 3n
+
+Final Answer:
+O(1 + 4n)
```

Comment:
When doing runtime complexity you always want to simplify the result which means ignoring constants and multipliers. So `O(1 + 4n)` could be simplified to `O(n)` which is linear. Here's a video I made of how to verify runtime complexities by counting the number of iterations a function goes through: https://youtu.be/0Z7yIEd8fJc